### DISCUSSION TITLE
export_table with gt? flextable? other?

### DISCUSSION THREAD

**Discussion by mattansb** [easystats/insight#263] [what's your opinion]
This is very easy to implement:

``` r
library(bayestestR)
#> Note: The default CI width (currently `ci=0.89`) might change in future versions (see https://github.com/easystats/bayestestR/issues/250). To prevent any issues, please set it explicitly when using bayestestR functions, via the 'ci' argument.
library(magrittr)
library(report)
#> report is in alpha - help us improve by reporting bugs on github.com/easystats/report/issues
library(insight)

lm1 <- lm(Sepal.Length ~ 1, data = iris)
lm2 <- lm(Sepal.Length ~ Species, data = iris)
lm3 <- lm(Sepal.Length ~ Species + Petal.Length, data = iris)
lm4 <- lm(Sepal.Length ~ Species * Petal.Length, data = iris)


t <- bayesfactor_models(lm1, lm2, lm3, lm4, denominator = 1) %>% 
  report_table() %>% 
  format()

attr(t, "table_caption") <- "Bayesian Model Comparison"
attr(t, "table_subtitle") <- "\nSome more words..."

cat(export_table(t))
#> Bayesian Model Comparison 
#> Some more words...
#> 
#> Model                      | BF against model 1 | BF against model 3 (best model)
#> ---------------------------------------------------------------------------------
#> [1] (Intercept only)       |               1.00 |                      1/5.84e+55
#> [2] Species                |           1.70e+29 |                      1/3.45e+26
#> [3] Species + Petal.Length |           5.84e+55 |                            1.00
#> [4] Species * Petal.Length |           2.20e+54 |                         1/26.52
#> 
#> Bayes Factor Type: BIC approximation
#> Against denominator: Model 1

export_table(t, format = "markdown")
```

<table>
<caption>Bayesian Model Comparison
Some more words…</caption>
<thead>
<tr class="header">
<th style="text-align: left;">Model</th>
<th style="text-align: right;">BF against model 1</th>
<th style="text-align: right;">BF against model 3 (best model)</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">[1] (Intercept only)</td>
<td style="text-align: right;">1.00</td>
<td style="text-align: right;">1/5.84e+55</td>
</tr>
<tr class="even">
<td style="text-align: left;">[2] Species</td>
<td style="text-align: right;">1.70e+29</td>
<td style="text-align: right;">1/3.45e+26</td>
</tr>
<tr class="odd">
<td style="text-align: left;">[3] Species + Petal.Length</td>
<td style="text-align: right;">5.84e+55</td>
<td style="text-align: right;">1.00</td>
</tr>
<tr class="even">
<td style="text-align: left;">[4] Species * Petal.Length</td>
<td style="text-align: right;">2.20e+54</td>
<td style="text-align: right;">1/26.52</td>
</tr>
</tbody>
</table>

Bayesian Model Comparison
Some more words…

Bayes Factor Type: BIC approximation

```R
library(gt)
gt(t) %>% 
  tab_header(
    title = attr(t, "table_caption"),
    subtitle = attr(t, "table_subtitle")
  ) %>% 
  tab_source_note(attr(t, "table_footer"))
```
![image](https://user-images.githubusercontent.com/35330040/101903632-422c7680-3bbd-11eb-9ff3-c29c2be0cfb5.png)

**Comment by strengejacke** [comment]
I think we can discuss some pros and cons here, @mattansb @DominiqueMakowski @IndrajeetPatil and others? There are a lot of other packages, and I think I also mentioned _flextable_ sometime, somewhere...

@mattansb Does your example render to markdown, or PDF/HTML?

**Reply by mattansb** [reply]
It renders to html, so can be exported to:
- html - yes
- pdf - yes
- word - no :(

**Reply by mattansb** [reply]
However, the implementation is so easy, I see no reason not to add it ^_^
(Or at least as an example is `parameters_table()`?)

**Reply by strengejacke** [reply]
> However, the implementation is so easy, I see no reason not to add it ^_^

True. Only reason could be, if we have multiple options to create HTML tables, we either need to decide _now_ for the best package to use, or to have an option/argument to set the "engine".

**Reply by mattansb** [reply]
Isn't that what `format` argument does?

**Reply by strengejacke** [reply]
No, `format()` actually call `insight::parameters_table()`, to create these fancy CI-columns, add pretty names as parameter names, etc. - and maybe some other preparation like adding title/caption/footer attributes. `format()` just returns a "plain" data frame.

However, the style of table "syntax" is done in `export_table()` (like plain text, or these `:---` for markdown table lines etc. `export_table()` create the character vector with markdown syntax, HTML syntax...

**Reply by mattansb** [reply]
I meant the `format` argument in `export_table()` (:

**Reply by strengejacke** [reply]
Ok, got it. 

> Isn't that what format argument does?

No, that is what I meant. `format` tells the function whether to use plain text, markdown, html, ... but not, which package to use to create the table string in markdown, html, ... So either, we decide now for a package, or we need to include the "engine", e.g. `format = "html_gt"` or `format = "html_flextable"`.

**Reply by mattansb** [reply]
Okay, now I understand. I like the latter option more...

**Comment by strengejacke** [comment]
> This is very easy to implement:

The question is, will it remain that easy for more complex models?

``` r
library(insight)
library(parameters)
m <- download_model("brms_zi_2")
mp <- model_parameters(m, effects = "all", component = "all")
print_md(mp)
```

<table>
<caption># Fixed effects (conditional)</caption>
<thead>
<tr class="header">
<th style="text-align: left;">Parameter</th>
<th style="text-align: right;">Median</th>
<th style="text-align: left;">CI</th>
<th style="text-align: left;">pd</th>
<th style="text-align: right;">% in ROPE</th>
<th style="text-align: left;">Rhat</th>
<th style="text-align: left;">ESS</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">(Intercept)</td>
<td style="text-align: right;">-0.86</td>
<td style="text-align: left;">(-1.37, -0.34)</td>
<td style="text-align: left;">98.22%</td>
<td style="text-align: right;">1.30%</td>
<td style="text-align: left;">1.002</td>
<td style="text-align: left;">1003</td>
</tr>
<tr class="even">
<td style="text-align: left;">persons</td>
<td style="text-align: right;">0.84</td>
<td style="text-align: left;">(0.68, 1.01)</td>
<td style="text-align: left;">100%</td>
<td style="text-align: right;">0%</td>
<td style="text-align: left;">1.002</td>
<td style="text-align: left;">878</td>
</tr>
<tr class="odd">
<td style="text-align: left;">child</td>
<td style="text-align: right;">-1.15</td>
<td style="text-align: left;">(-1.29, -1.00)</td>
<td style="text-align: left;">100%</td>
<td style="text-align: right;">0%</td>
<td style="text-align: left;">1.001</td>
<td style="text-align: left;">2166</td>
</tr>
<tr class="even">
<td style="text-align: left;">camper</td>
<td style="text-align: right;">0.73</td>
<td style="text-align: left;">(0.58, 0.89)</td>
<td style="text-align: left;">100%</td>
<td style="text-align: right;">0%</td>
<td style="text-align: left;">1.000</td>
<td style="text-align: left;">3528</td>
</tr>
</tbody>
</table>

<table>
<caption># Fixed effects (zero-inflated)</caption>
<thead>
<tr class="header">
<th style="text-align: left;">Parameter</th>
<th style="text-align: right;">Median</th>
<th style="text-align: right;">CI</th>
<th style="text-align: left;">pd</th>
<th style="text-align: right;">% in ROPE</th>
<th style="text-align: left;">Rhat</th>
<th style="text-align: left;">ESS</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">(Intercept)</td>
<td style="text-align: right;">-0.65</td>
<td style="text-align: right;">(-1.82, 0.47)</td>
<td style="text-align: left;">83.17%</td>
<td style="text-align: right;">7.80%</td>
<td style="text-align: left;">1.000</td>
<td style="text-align: left;">1174</td>
</tr>
<tr class="even">
<td style="text-align: left;">child</td>
<td style="text-align: right;">1.86</td>
<td style="text-align: right;">(1.37, 2.42)</td>
<td style="text-align: left;">100%</td>
<td style="text-align: right;">0%</td>
<td style="text-align: left;">1.002</td>
<td style="text-align: left;">3022</td>
</tr>
<tr class="odd">
<td style="text-align: left;">camper</td>
<td style="text-align: right;">-0.84</td>
<td style="text-align: right;">(-1.40, -0.24)</td>
<td style="text-align: left;">98.95%</td>
<td style="text-align: right;">1.68%</td>
<td style="text-align: left;">1.000</td>
<td style="text-align: left;">3410</td>
</tr>
</tbody>
</table>

<table>
<caption># Random effects (conditional) SD/Cor: persons</caption>
<thead>
<tr class="header">
<th style="text-align: left;">Parameter</th>
<th style="text-align: right;">Median</th>
<th style="text-align: left;">CI</th>
<th style="text-align: left;">pd</th>
<th style="text-align: right;">% in ROPE</th>
<th style="text-align: left;">Rhat</th>
<th style="text-align: left;">ESS</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">(Intercept)</td>
<td style="text-align: right;">0.12</td>
<td style="text-align: left;">(1.65e-04, 0.39)</td>
<td style="text-align: left;">100%</td>
<td style="text-align: right;">45.27%</td>
<td style="text-align: left;">1.005</td>
<td style="text-align: left;">819</td>
</tr>
</tbody>
</table>

<table>
<caption># Random effects (zero-inflated) SD/Cor: persons</caption>
<thead>
<tr class="header">
<th style="text-align: left;">Parameter</th>
<th style="text-align: right;">Median</th>
<th style="text-align: left;">CI</th>
<th style="text-align: left;">pd</th>
<th style="text-align: right;">% in ROPE</th>
<th style="text-align: left;">Rhat</th>
<th style="text-align: left;">ESS</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">(Intercept)</td>
<td style="text-align: right;">1.32</td>
<td style="text-align: left;">(0.48, 2.49)</td>
<td style="text-align: left;">100%</td>
<td style="text-align: right;">0%</td>
<td style="text-align: left;">1.004</td>
<td style="text-align: left;">1331</td>
</tr>
</tbody>
</table>

<sup>Created on 2020-12-11 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

So, how to either a) insert "header rows", or b) append multiple table objects? I just mean how to do this in `gt()`, the implementation in `export_table()` should be easy.

**Reply by mattansb** [reply]
```R
library(insight)
library(parameters)
m <- download_model("brms_zi_2")
mp <- model_parameters(m, effects = "all", component = "all")

library(gt)

mp %>% 
  dplyr::group_by(Component, Effects) %>% 
  gt()
```

![image](https://user-images.githubusercontent.com/35330040/101947068-8ab55580-3bf8-11eb-9393-d2bc9269c0b9.png)

(Couldn't copy the whole thing... But you get the idea)

**Reply by strengejacke** [reply]
Ok, there's a `groupname_col`-argument that does the trick:

``` r
library(insight)
library(parameters)
m <- download_model("brms_zi_2")
mp <- model_parameters(m, effects = "all", component = "all")

library(gt)
gt(mp, groupname_col = c("Effects", "Component"))
```

**Comment by strengejacke** [comment]
OMG, @strengejacke 

![l75h5qfiy4l01](https://user-images.githubusercontent.com/26301769/101960797-7d07cc00-3c08-11eb-83e7-9b335d31d334.jpg)


Yes...

https://rpubs.com/sjPlot/export_table

**Reply by strengejacke** [reply]
`export_table(format = "html") needs some improvements, though:

- use table title for html
- currently, subtables are split by columns `Response`, `Effects` and `Component` - we may add an argument here for user-defined split-columns.

**Reply by mattansb** [reply]
Awesome!!

> currently, subtables are split by columns Response, Effects and Component - we may add an argument here for user-defined split-columns.

Shouldn't group also be included?

``` r
df <- iris
df$Sepal.Big <- ifelse(df$Sepal.Width >= 3, "Yes", "No")

model <- aov(Sepal.Length ~ Sepal.Big + Error(Species), data = df)

parameters::model_parameters(model)
#> # Species
#> 
#> Parameter | Sum_Squares | df | Mean_Square |    F |     p
#> ---------------------------------------------------------
#> Sepal.Big |       28.27 |  1 |       28.27 | 0.81 | 0.534
#> Residuals |       34.94 |  1 |       34.94 |      |      
#> 
#> # Within
#> 
#> Parameter | Sum_Squares |  df | Mean_Square |     F |      p
#> ------------------------------------------------------------
#> Sepal.Big |        4.74 |   1 |        4.74 | 20.24 | < .001
#> Residuals |       34.21 | 146 |        0.23 |       |
```

<sup>Created on 2020-12-13 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>
